File: Basic/Java/Day-11/CircularQueue.java
public class CircularQueue {

    private int[] queue;
    private int capacity;
    private int front = 0;
    private int rear = -1;
    private int size = 0;

    public CircularQueue(int capacity) {
        this.capacity = capacity;
        this.queue = new int[capacity];
    }

    public void enqueue(int value) {
        if (size == capacity) {
            System.out.println("Queue is full!");
            return;
        }
        rear = (rear + 1) % capacity;
        queue[rear] = value;
        size++;
    }

    public int dequeue() {
        if (size == 0) {
            System.out.println("Queue is empty!");
            return -1;
        }
        int value = queue[front];
        front = (front + 1) % capacity;
        size--;
        return value;
    }

    public int peek() {
        if (size == 0) {
            System.out.println("Queue is empty!");
            return -1;
        }
        return queue[front];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public int getSize() {
        return size;
    }

    public void display() {
        if (size == 0) {
            System.out.println("Queue is empty!");
            return;
        }
        System.out.print("Queue elements: ");
        for (int i = 0; i < size; i++) {
            System.out.print(queue[(front + i) % capacity] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int capacity = 5;
        CircularQueue cq = new CircularQueue(capacity);

        cq.enqueue(10);
        cq.enqueue(20);
        cq.enqueue(30);
        cq.enqueue(40);
        cq.enqueue(50);
        cq.display();

        System.out.println("Dequeued: " + cq.dequeue());
        System.out.println("Dequeued: " + cq.dequeue());
        cq.display();

        cq.enqueue(60);
        cq.enqueue(70);
        cq.display();

        System.out.println("Front element: " + cq.peek());
        System.out.println("Size: " + cq.getSize());
        System.out.println("Is Full: " + cq.isFull());
        System.out.println("Is Empty: " + cq.isEmpty());
    }
}

File: BasicToAdvance/Python/Day-11/CircularQueue.py
class CircularQueue:

    def __init__(self,capacity):
        self.queue = [None]* capacity
        self.capacity = capacity
        self.front = 0
        self.rear = -1
        self.size = 0

    def enqueue(self, value):
        if self.is_full():
            return False

        self.rear = (self.rear + 1)% self.capacity
        self.queue[self.rear]=value
        self.size += 1

        return True

    def dequeue(self):
        if self.is_empty():
            return None

        value = self.queue[self.front]
        self.queue[self.front]= None
        self.front = (self.front+1)% self.capacity
        self.size -= 1

        return value

    def peek(self):
        if self.is_empty():
            return None

        return self.queue[self.front]

    def is_empty(self):
        return self.size == 0

    def Size(self):
        return self.size

    def is_full(self):
        return self.size == self.capacity

    def show(self):
        return self.queue

    def empty_slots(self):
        if self.is_full():
            return False

        print(f"Yes...It has {self.capacity-self.size} empty slot.")


capacity = 5
cq = CircularQueue(capacity)

cq.enqueue(10)
cq.enqueue(20)
cq.enqueue(30)
cq.enqueue(40)
cq.enqueue(50)

print(cq.show())
print(cq.dequeue())
print(f"Size of the Circular Queue is: {cq.Size()}")
print(cq.show())
cq.empty_slots()

print(f"Enqueue operation is succesfull ? : {cq.enqueue(60)}")
print(cq.show())
print(f"Enqueue operation is succesfull ? : {cq.enqueue(70)}")

print(f"Size of the Circular Queue is: {cq.Size()}")
print(f"Circular queue is full : {cq.is_full()}")
print(f"Circular queue is empty : {cq.is_empty()}")